File: 0029-divide-two-integers/0029-divide-two-integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        if (dividend==Integer.MIN_VALUE&&divisor==-1)return Integer.MAX_VALUE;
        boolean negative=(dividend<0)^(divisor<0);
        long dvd=Math.abs((long)dividend);
        long dvs=Math.abs((long)divisor);
        long quotient=0;
        for (int i=31;i>=0;i--){
            if ((dvs<<i)<=dvd){
                dvd-=(dvs<<i);
                quotient+=(1L<<i);
            }
        }
        return negative?(int)-quotient:(int)quotient;
    }
}

File: 0029-divide-two-integers/0029-divide-two-integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend==INT_MIN && divisor==-1) return INT_MAX;
        bool neg=(dividend<0)^(divisor<0);
        long long dvd=llabs((long long)dividend);
        long long dvs=llabs((long long)divisor);
        long long ans=0;
        for(int i=31;i>=0;i--){
            if ((dvs<<i)<=dvd){
                dvd-=(dvs<<i);
                ans+=(1LL<<i);
            }
        }
        return neg?-ans:ans;
    }
};